### PR TITLE
Stop ignoring declared types when initializing expressions are tuples

### LINKED ISSUE #18219: init= not called with a isTuple() where clause
Currently we allow `init=` for records where the RHS does not need to be the same type as the Record itself. I was trying to utilize tuples in an `init=` function and was encountering a situation where it seemed like the function was not being called at all.

Here is a small reproducer of the same issue

A trivial record with two values
```chapel
record Foo {
    var x: int;
    var y: string;

    proc init(x: int, y: string){
        this.x = x + 1;
        this.y = y + "!";
    }

    proc init=(tup) where isTuple(tup){
        this.init(tup(0), tup(1));
    }

}
```

I initialized two variables here, one using the `init` function and one using  (in theory) `init=`
```chapel
var bar: Foo = new Foo(10, "Hi"); // Should call init
var baz: Foo = (11, "Bye"); // Should call init=
```

But when I printed these variables and some data associated with them I got the following:
```chapel
writeln(bar);
writeln(baz); // Quietly changed baz's type from Foo to a normal Tuple?

writeln(isTuple(bar)); // false
writeln(isTuple(baz)); // true ??
```
outputs:

```bash
(x = 11, y = Hi!)
(11, Bye)
false
true
```
This behavior seems very buggy since the correct function isn't called and the type of a variable is changed from what it was explicitly typed to be with no indication that it happened.

### PR DESCRIPTION
**Pull request by bradcray**
This fixes a bug in our code base that Shreyas ran into, in which if a
variable's initializer is a tuple:

```chapel
var r: T = (t1, t2);
```

then the type of the variable 'T' is completely ignored and 'r' is
treated as a tuple, as though it had been declared:

```chapel
var r = (t1, t2);
```

The effect was that if `T` was a record with a copy initializer
accepting tuples, say, the type of `T` would be dropped on
the floor and the copy initializer never called.

It turns out that this was because when resolution found any
declaration initialized with a tuple, it rewrote the declaration
essentially as:

```chapel
var r = initCopy((t1, t2));
```
Here, I only make this initCopy() rule fire if both the rhs and
lhs are tuple types, which is presumably what was intended.

This is one of those classic fixes where part of me thinks
"Are there other things we should be checking for before
applying this rule to tuples?" (like "same size?" "same element
type?") and/or "Are there other types in this conditional that
are similarly causing declared types to be ignored?" but in
the interest of not getting too distracted from my main tasks,
I'm considering this an improvement that doesn't break anything
in our testing system and leaving those questions open for
another day.

Resolves #18219.

### CODE REVIEW

**Comment by bradcray** [comment]
Tagging @ShreyasKhandekar on this since he filed the bug.

File: compiler/resolution/functionResolution.cpp [hunk 1]
```diff
@@ -7110,7 +7110,7 @@ void resolveInitVar(CallExpr* call) {
              targetType->getValType()->symbol->hasFlag(FLAG_ARRAY) ||
              isDomainWithoutNew ||
              initCopySyncSingle ||
-             srcType->getValType()->symbol->hasFlag(FLAG_TUPLE) ||
+             (targetType->getValType()->symbol->hasFlag(FLAG_TUPLE) && srcType->getValType()->symbol->hasFlag(FLAG_TUPLE)) ||
```
Comment (lydia-duncan) [inline compiler/resolution/functionResolution.cpp:7113]: This looks too long for an 80 character line - I think you can split it on the `&&`

**Review by lydia-duncan — APPROVED** [review]
Looks good to me, modulo the minor line length fix

File: compiler/resolution/functionResolution.cpp (hunk 1 above)
Comment (bradcray) [reply compiler/resolution/functionResolution.cpp]: :+1:

**Pull request merged**